%\VignetteIndexEntry{Introduction to QDNAseq}

\documentclass{article}

\begin{document}

\title{Introduction to QDNAseq}
\author{Ilari Scheinin}
\maketitle

\section{Running QDNAseq}

First load the package.

<<reg>>=
library(QDNAseq)
@

Next we need the bin annotations. These are available pre-calculated for genome build hg19 and bin sizes 1, 5, 10, 15, 30, 50, 100, 500, and 1000 kbp. They can be loaded for example with:

<<reg, eval=FALSE>>=
bins <- downloadBinAnnotations(binSize=15)
@

If you are working with another genome build (or another species), see the section on generating the bin annotations.

To load all the bam files in the current directory, run:

<<reg, echo=TRUE, eval=FALSE>>=
readCounts <- binReadCounts(bins)
# all files ending in .bam from the current working directory

# or

readCounts <- binReadCounts(bins, bamfiles='tumor.bam')
# file 'tumor.bam' from the current working directory

# or

readCounts <- binReadCounts(bins, path='tumors')
# all files ending in .bam from the subdirectory 'tumors'
@

For the purpose of this tutorial, we load the example data set of chromosome 9 of low grade glioma sample LGG150.

<<reg>>=
data(LGG150)
@

Plot a raw copy number profile (read counts across the genome), and highlight bins removed with default filtering.

<<label=rawprofile,include=FALSE>>=
plot(LGG150, ylim=c(-50, 150))
highlightFilters(LGG150, residual=TRUE, blacklist=TRUE)
@

\begin{figure}
\begin{center}
<<label=rawprofile,fig=TRUE,echo=FALSE>>=
<<rawprofile>>
@
\end{center}
\caption{A plot}
\label{fig:rawprofile}
\end{figure}

Apply filters and plot median read counts as a function of GC content and mappability. As the example data set only contains chromosome 9 and not the whole genome, this looks jagged. Normally, a smoother distribution is to be expected.

<<label=readcounts,include=FALSE>>=
LGG150 <- applyFilters(LGG150, residual=TRUE, blacklist=TRUE)
isobarPlot(LGG150)
@

\begin{figure}
\begin{center}
<<label=readcounts,fig=TRUE,echo=FALSE>>=
<<readcounts>>
@
\end{center}
\caption{A plot}
\label{fig:readcounts}
\end{figure}

Perform correction for GC content and mappability, normalize, and plot the copy number profile.

<<reg>>=
LGG150 <- correctBins(LGG150)
LGG150 <- normalizeBins(LGG150)
@

<<label=profile,include=FALSE>>=
plot(LGG150)
@

\begin{figure}
\begin{center}
<<label=profile,fig=TRUE,echo=FALSE>>=
<<profile>>
@
\end{center}
\caption{A plot}
\label{fig:profile}
\end{figure}

Data is now ready to be analyzed with a downstream package of choice. Segmentation with the CBS algorithm from DNAcopy and calling copy number aberrations with CGHcall have been implemented for convenience.

<<reg>>=
LGG150 <- segmentBins(LGG150)
@

Plot to evaluate segmentation.

<<label=segmentation,include=FALSE>>=
plot(LGG150)
@

\begin{figure}
\begin{center}
<<label=segmentation,fig=TRUE,echo=FALSE>>=
<<segmentation>>
@
\end{center}
\caption{A plot}
\label{fig:segmentation}
\end{figure}

Tune segmentation parameters and iterate until satisfied. Next, call aberrations, and plot the final results.

<<reg>>=
# LGG150 <- callBins(LGG150)
@

<<label=calls,include=FALSE>>=
plot(LGG150)
@

\begin{figure}
\begin{center}
<<label=calls,fig=TRUE,echo=FALSE>>=
<<calls>>
@
\end{center}
\caption{A plot}
\label{fig:calls}
\end{figure}

Finally, for other downstream analyses, such as running CGHregions, it might be useful to convert to a cghCall object.

<<reg>>=
# cgh <- makeCgh(LGG150)
@

% caching

% 661 bam files = 445G
% cache/reads = 13G
% cache/readCounts/15kbp = 139M
% cache/readCounts/30kbp = 82M
% cache/readCounts/100kbp = 32M
% cache/readCounts/1000kbp = 5.2M

\section{Generating bin annotations}

<<reg, eval=FALSE>>=
# load required packages for human reference genome build hg19
library(QDNAseq)
library(Biobase)
library(BSgenome.Hsapiens.UCSC.hg19)

# set the bin size
binSize <- 15

# create bins from the reference genome
bins <- createBins(bsgenome=BSgenome.Hsapiens.UCSC.hg19, binSize=binSize)

# calculate mappabilites per bin from ENCODE mapability tracks
bins$mappability <- calculateMappability(bins,
  bigWigFile='/path/to/wgEncodeCrgMapabilityAlign50mer.bigWig',
  bigWigAverageOverBed='/path/to/bigWigAverageOverBed')

# calculate overlap with ENCODE blacklisted regions
bins$blacklist <- calculateBlacklist(bins,
  bedFiles=c('/path/to/wgEncodeDacMapabilityConsensusExcludable.bed',
  '/path/to/wgEncodeDukeMapabilityRegionsExcludable.bed'))

# load data for the 1000 Genomes (or similar) data set, and generate residuals
tg <- binReadCounts(bins,
  path='/path/to/1000Genomes/bam/files')
tg <- applyFilters(tg, residual=FALSE, blacklist=FALSE,
  mappability=FALSE, bases=FALSE)
bins$residual <- iterateResiduals(tg)

# by default, use all autosomal bins that have a reference sequence
# (i.e. not only N's)
bins$use <- bins$chromosome %in% as.character(1:22) & bins$bases > 0

# convert to AnnotatedDataFrame and add metadata
bins <- AnnotatedDataFrame(bins,
  varMetadata=data.frame(labelDescription=c(
  'Chromosome name',
  'Base pair start position',
  'Base pair end position',
  'Percentage of non-N nucleotides (of full bin size)',
  'Percentage of C and G nucleotides (of non-N nucleotides)',
  'Average mappability of 50mers with a maximum of 2 mismatches',
  'Percent overlap with ENCODE blacklisted regions',
  'Median loess residual from 1000 Genomes (50mers)',
  'Whether the bin should be used in subsequent analysis steps'),
  row.names=colnames(bins)))

# add more metadata
attr(bins, 'QDNAseqVersion') <- packageVersion('QDNAseq')
attr(bins, 'QDNAseqDownloaded') <- TRUE
saveRDS(bins, paste('QDNAseq.hg19.', binsize, 'kbp.SR50.rds', sep=''),
  compress='xz')
@

% library(QDNAseq)
% for (binSize in c(1, 5, 10, 15, 30, 50, 100, 500, 1000)) {
%   bins <- downloadBinAnnotations(binSize)
%   tg <- binReadCounts(bins,
%     path='/storage/shark1/data/SeqAnalysis/1000Genomes/read.length.50/bam/')
%     path='/storage/shark1/data/SeqProjects/1000G.50.hg19/bam/')
%   tg <- correctBins(tg, storeResiduals=TRUE)
%   residual <- apply(Biobase::assayDataElement(tg, 'residuals'), 1,
%     median, na.rm=TRUE)
%   residual[!bins$chromosome %in% as.character(1:22)] <- NA
%   bins$residual <- residual
%   attr(bins, 'residualMadDiff') <-
%     matrixStats::madDiff(bins$residual, na.rm=TRUE)
%   attr(bins, 'QDNAseqVersion') <- packageVersion('QDNAseq')
%   attr(bins, 'QDNAseqDownloaded') <- TRUE
%   saveRDS(bins, paste('QDNAseq.hg19.', binSize, 'kbp.rds', sep=''),
%     compress='xz')
% }

% where to download ENCODE
% how 1000 genomes were downloaded

\end{document}

% EOF
